:recycle: Refactor cache system

ashwinvin opened the pull request:
## What does this PR do?
This PR implements the following things:
- [x] Refactor the cache system to make it more testable and extensible.
- [x] Fix search results not being cached.

## Why is this change important?
The current cache system is limited and is tightly coupled to just caching search results.
This PR abstracts the main logic and opens it up for newer cache backends which could be implemented.

## How to test this PR locally?
`cargo test`

## Author's Additional Notes
Should admins be allowed to set the max number of search results to cache and TTL of search results?

## Related issues
Closes #382

ashwinvin committed:
:recycle: Refactor cache system

neon-mmd commented:
The changes so far looks great :+1:  I think the trait based approach would be better because I see your point having this approach would be more extensible. So this a completely `yes` from my side. :slightly_smiling_face: 

---

> Should admins be allowed to set the max number of search results to cache and TTL of search results?

Also, about how many results could be stored in In-Memory caching solution and the expiry of results in both the caching solutions. I think it would be better to provide control for this to the admin as the admin would have control on how much memory to allocate as well as for how long cache should remain valid. For these options we could provide to options in the config and for this work I think we can have a separate issue for this. :slightly_smiling_face:

ashwinvin committed:
:bug: Fix cache not getting set

This patch also makes it that cookies are eagerly evaluated. This is
done to figure out the safe search level set by the user. The
performance hit wouldn't be much of a deal as the cookie is a small
json string

ashwinvin commented:
Its pretty easy to add the configuration settings for that. Yeah, lets put up a new issue to add those settings once this is merged.

alamin655 committed:
🔖 chore: bump the app version (#399)

alamin655 committed:
🔖 chore: bump the app version (#399)

neon-mmd reviewed (approved):
Thanks :heart: for working on such a wonderful PR. We really appreciate it :+1: Keep it up!! Since everything looks good. We will merge this PR right away :rocket: :slightly_smiling_face: .
